internal/provider/privateinternetaccess/wireguardconfigurator.go

```diff
@@ -0,0 +1,104 @@
+package privateinternetaccess
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/qdm12/gluetun/internal/configuration/settings"
+	"github.com/qdm12/gluetun/internal/models"
+	"github.com/qdm12/gluetun/internal/provider/utils"
+	"github.com/qdm12/gluetun/internal/wireguard"
+	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+	"net"
+	"net/http"
+	"net/netip"
+	"net/url"
+	"strconv"
+)
+
+func (p *Provider) GetWireguardConnection(ctx context.Context, connection models.Connection, wireguardSettings settings.Wireguard, ipv6Supported bool) (settings wireguard.Settings, err error) {
+	// create http client to make requests to the server's API
+	client, err := newHTTPClient(connection.Hostname)
+
+	// generate a new private key
+	privateKey, err := wgtypes.GeneratePrivateKey()
+
+	// generate public key from private key
+	publicKey := privateKey.PublicKey()
+
+	// fetch token from PIA's API
+	token, err := fetchToken(ctx, client, "gtoken", p.authFilePath)
```

Comment:
⚠️ This calls www.privateinternetaccess.com and that's a deal breaker for the killswitch and to avoid leaking data outside the tunnel (especially from other containers/devices connecting through gluetun). For example if the tunnel takes 5 seconds to setup, data will leak for 5 seconds; if the tunnel never manages to setup, it would leak for hours.
That's also why I did not implement this yet, but don't worry, your code will be used at some point.

I'm working on a few other moving parts, and the idea is that connecting to privateinternetaccess.com over https (and its dns resolution) would be the only one allowed outside the vpn, temporarily. But this is far from done, so expect this pull request to stay opened for a few months unfortunately.

Reply:
Theoretically, if the agent were to connect via openvpn via a sideband, download the necessary wireguard configs over the ovpn tunnel, then bring up the primary wireguard tunnel, would that address the leakage risk?

Obviously the scenario assumes all tunnelled traffic is only routed over wireguard, and thus is blocked till the wireguard tunnel is up. I would assume this should be the default behaviour anyways in the event of the PIA dropping the wireguard tunnel to rotate IPs.